File: Tree/Binary Tree/Diameteroftree.cpp
  #include<queue>

      int height(Node* node){
      
      
      if(node==NULL){
            return 0;
      }
      
      int left=height(node->left);
      int right=height(node->right);
      
      if(left>right){
      return left+1;
      }
      
      else{
            return right+1;
      }
      
      
      }

      int diameter(Node* node) {
      int max=INT_MIN;
      queue<Node*>q;
      
      q.push(node);
      
      while(q.size()>0){
            
            Node*temp=q.front();
            q.pop();
            
            int tl=0;
            int tr=0;
            int p=0;
            
            
      // if(temp->left!=NULL && temp->right!=NULL){
            
            tl=height(temp->left);
            tr=height(temp->right);
            
            // cout<<"here"<<endl;
            
            p=tl+tr+1;
            
            if(p>max)
            {
                  max=p;
            }
      // }
            
      
            
      
      
            
            if(temp->left!=NULL)
            {
            
            q.push(temp->left);
                  
                  
            }
            
            if(temp->right!=NULL)
            {
            q.push(temp->right);
            }
            
            
      }
      
            return max;

      
      
      }

File: Graphs/Diameter_of_tree/Diameter_of_tree.java
class Node {
	int data;
	Node left, right;

	public Node(int item)
	{
		data = item;
		left = right = null;
	}
}
class BinaryTree {
	Node root;
	int diameter(Node root)
	{
		if (root == null)
			return 0;
		int lheight = height(root.left);
		int rheight = height(root.right);
		int ldiameter = diameter(root.left);
		int rdiameter = diameter(root.right);
    return Math.max(lheight + rheight + 1,
						Math.max(ldiameter, rdiameter));
	}
	int diameter() { return diameter(root); }
	static int height(Node node)
	{
		if (node == null)
			return 0;
		return (1
				+ Math.max(height(node.left),
						height(node.right)));
	}
	public static void main(String args[])
	{
		// creating a binary tree and entering the nodes
		BinaryTree tree = new BinaryTree();
		tree.root = new Node(1);
		tree.root.left = new Node(2);
		tree.root.right = new Node(3);
		tree.root.left.left = new Node(4);
		tree.root.left.right = new Node(5);
    
		System.out.println(
			"The diameter of given binary tree is : "
			+ tree.diameter());
	}
}
